### PR TITLE
Allow dynamic user variables to be used with filter rules

### PR DESCRIPTION
**Pull request by licitdev**
This feature addresses #2993 by allowing users to use other columns from the `directus_users` table when doing filtering in permissions or conditions. Environment variable is the chosen method as it reduces the need for another database query.
 
This feature could be beneficial for those working on SaaS products such as in #7054 and #7236 where they can filter by users' organisation or group fields.

### Usage
1. Add the list of column names to the `USER_DYNAMIC_VARS` environment variable, similar to how the `DB_EXCLUDE_TABLES` works. Example: `USER_DYNAMIC_VARS="email,first_name,tags"`.
2. Use `$CURRENT_USER.<column name>` when filtering in permissions or conditions.

### Todo
1. Fix the preview of filtering conditions automatically replacing the variables such as `$CURRENT_USER` and `$CURRENT_USER.email`. It is an issue if the admin views the rule and save it after, which causes the dynamic variable to be saved with the admin's data.  NOTE: Opened issue #7894 
2. Update documentation to include usage of this feature.

### CODE REVIEW

**Comment by licitdev** [comment]
Hello @nickrum, I am not sure where to look to fix the issue in my Todo above.

I believe that the issue is with the parsing of field information when it is saved into the state (store). This issue only occurs for conditions, this issue does not exist for item filter rules in permissions.

The filter rule is stored correctly in the database as follows
```
[
  {
    "name": "matches email",
    "rule": {
      "email": {
        "_neq": "$CURRENT_USER.email"
      }
    },
    "hidden": false,
    "readonly": true
  }
]
```

But when displayed in `app/src/modules/settings/routes/data-model/field-detail/components/conditions.vue`, the store contains the following data, with the `$CURRENT_USER.email` in the conditions automatically replaced with `user@example.com`.

```
{
    "field": "special",
    "type": "string",
    "schema": {
        "name": "special",
        "table": "testing",
        "data_type": "character varying",
        "default_value": null,
        "generation_expression": null,
        "max_length": 255,
        "numeric_precision": null,
        "numeric_scale": null,
        "is_generated": false,
        "is_nullable": true,
        "is_unique": false,
        "is_primary_key": false,
        "has_auto_increment": false,
        "comment": null,
        "schema": "public",
        "foreign_key_schema": null,
        "foreign_key_table": null,
        "foreign_key_column": null
    },
    "meta": {
        "id": 5,
        "collection": "testing",
        "field": "special",
        "special": null,
        "interface": null,
        "options": null,
        "display": null,
        "display_options": null,
        "readonly": false,
        "hidden": false,
        "sort": null,
        "width": "full",
        "group": null,
        "translations": null,
        "note": null,
        "conditions": [
            {
                "name": "matches email",
                "rule": {
                    "email": {
                        "_neq": "user@example.com"
                    }
                },
                "hidden": false,
                "readonly": true
            }
        ],
        "required": false
    }
}
```

**Comment by nickrum** [comment]
Hi @licitdev, I am not the original author of this part of the codebase, so I can't be of much help I'm afraid.

**Comment by licitdev** [comment]
Thank you @nickrum. No worries, I've managed to find the source and submitted fix #7907.

File: api/src/middleware/authenticate.ts [hunk 1]
```diff
@@ -39,7 +43,12 @@ const authenticate: RequestHandler = asyncHandler(async (req, res, next) => {
...
 
 		const user = await database
-			.select('role', 'directus_roles.admin_access', 'directus_roles.app_access')
+			.select(
+				'role',
+				'directus_roles.admin_access',
+				'directus_roles.app_access',
+				...userDynamicVars.map((v) => 'directus_users.' + v)
+			)
 			.from('directus_users')
```
Comment (rijkvanzanten) [inline api/src/middleware/authenticate.ts:52]: We might have to start relying on a UserService instance here so we can also reliably fetch nested relational fields coming from the user for the purposes of applying permissions 🤔

File: api/src/middleware/authenticate.ts [hunk 2]
```diff
@@ -76,6 +94,9 @@ const authenticate: RequestHandler = asyncHandler(async (req, res, next) => {
 		req.accountability.role = user.role;
 		req.accountability.admin = user.admin_access === true || user.admin_access == 1;
 		req.accountability.app = user.app_access === true || user.app_access == 1;
+		for (const column of userDynamicVars) {
+			req.accountability.userDynamicVars[column] = user[column];
+		}
```
Comment (rijkvanzanten) [inline api/src/middleware/authenticate.ts:99]: Wondering if we should combine this under a "userData" key together with the users ID 🤔

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]
Another bigger question is to when to fetch these data points exactly. As a matter of fact, I was planning on removing that user data fetching at this point in the logic in favor of just relying on the payload of the JWT (as that's what it's made to do in the first place), to speed up the API a bit more. However, we wouldn't want to save these custom user data points for permissions into the JWT payload neither, as that data is publicly readable. 

We might want to extract the permissions fetching from the getSchema logic into a separate getPermissions function in which we can parse the filters, dynamically fetch any user data that is needed to transform the dynamic vars, and use the output filters down the data flow of the API 🤔

**Comment by rijkvanzanten** [comment]
Thanks @licitdev! Just a quick heads up: I don't expect you to change or even understand what I'm talking about above 😄 They're mostly _notes to self_ out of the top of my head while figuring out where this fits best with the bigger picture in mind

**Comment by licitdev** [comment]
Thank you @rijkvanzanten. I was developing with the least changes required to get this feature out (really hope this will be launched soon for SaaS purpose, I just needed one variable for filtering 😁). 

> Another bigger question is to when to fetch these data points exactly. As a matter of fact, I was planning on removing that user data fetching at this point in the logic in favor of just relying on the payload of the JWT (as that's what it's made to do in the first place), to speed up the API a bit more. However, we wouldn't want to save these custom user data points for permissions into the JWT payload neither, as that data is publicly readable.

We should be making use of a cache rather than in JWT as the user data may be changed at any point of time. However, when using cache for dynamic variables, it may result in the memory running out especially if not using an external cache store... The invalidation of cache will definitely have to be implemented upon updates to the dynamic columns. 

> We might want to extract the permissions fetching from the getSchema logic into a separate getPermissions function in which we can parse the filters, dynamically fetch any user data that is needed to transform the dynamic vars, and use the output filters down the data flow of the API 🤔

I am not familiar of the entire API flow but I am certain that the number of database queries have to be kept to a minimum. I can see how fetching user data in the getPermissions function is useful as columns can be removed depending on their role. However, this may pose an issue if the column filtered is not readable by a user when using field conditions in the app.

File: api/src/middleware/authenticate.ts (hunk 1 above)
Comment (licitdev) [reply api/src/middleware/authenticate.ts:52]: Fetching nested relational fields sounds awesome, it opens up more flexibility for permissions. 👍🏼

File: api/src/middleware/authenticate.ts (hunk 2 above)
Comment (licitdev) [reply api/src/middleware/authenticate.ts:99]: Somehow your comment reminded me of image file name combined with its transformation hash.
It got me thinking about the key used when storing in cache (if implemented).
Something along the lines of `{{user_id}}-{{column_name}}`. 😁

**Comment by licitdev** [comment]
@rijkvanzanten I have just tested that this feature works with your new permission filtering interface.
I have updated the docs to include more details. 

Screenshot 2021-10-03 at 7 02 14 PM

There is no additional database query added by this feature. Only additional columns are retrieved from `directus_users` through an existing DB query.
Hence, there should be a minimal difference should the entire application flow is to be changed as compared with other dynamic variables such as `$NOW` and `$CURRENT_ROLE`.

You mentioned nested relational fields in the conversation above but that isn't needed for this feature to work.
However, it could be a future improvement to this feature should the application flow be changed so that nested relational fields are optimized to be fetched only when required.

On a side note, only the raw data within the `directus_users` table is needed as it is only used when filtering other tables. There is no issue whether the field is a relationship field or standard field.

All in all, this feature allows for some forms of SaaS to make use of Directus, just with a single `company` field for filtering, or even complex SaaS that requires multiple attributes such as `user_group` and `plan_type`.
Hope that this feature will make it into the first release of Directus. 

Thank you! 🙏🏼

**Comment by danieljwestman** [comment]
We are at least two enthusiasts who maybe willing to finance a prioritization for the merge of this PR.

What do you say @rijkvanzanten and @benhaynes?

**Comment by rijkvanzanten** [comment]
Thanks @licitdev! I've started a new branch to implement all the other tweaks that are required for this PR here: https://github.com/directus/directus/pull/9376. I'll close this in favor of continuing the implementation on that 👍🏻

**Pull request closed without merging**